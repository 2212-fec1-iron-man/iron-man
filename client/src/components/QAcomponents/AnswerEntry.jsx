import { useState, useEffect} from 'react';
import ReactModal from 'react-modal';
import axios from 'axios';
import EnlargedImageModal from './EnlargedImageModal.jsx';

const AnswerEntry = ({ answer }) => {
    // console.log('what is each answer_id looks like :', answer.id);
    const [answerHelpfulCount, setAnswerHelpfulCount] = useState(answer.helpfulness);
    const [isOpen, setIsOpen] = useState(false);
    const [isEnlarged, setIsEnlarged] = useState(false);
    const [enlargePhotoIndex, setEnlargedPhotoIndex] = useState(-1);

    const [votedHelpful, setVotedHelpful] = useState(false);
    const [reportActed, setReportActed] = useState(false);

    const [reportReason, setReportReason] = useState("");
    const [wordCount, setWordCount] = useState(0);

    useEffect(() => {
      setWordCount(reportReason.split(" ").length);
    }, [reportReason]);

    const handleVote = (answerId) => {
        // console.log("am in handleVote func, prepare send request to API", apiUrl);
        axios.put(`/qa/answers/${answerId}/helpful`)
        .then((response) => {
          // alert("Thanks for voting this answer helpful! ")
          // console.log("voting succeed!")
          setVotedHelpful(true);
        })
        .catch ((error) => {
          // alert('this answer voting get error: ', error);
        })
    }

    const handleHelpfulClick = (answerId) => {
        setAnswerHelpfulCount(answerHelpfulCount + 1);
        handleVote(answerId);
    }

    const handleOpenModal = () => {
      setIsOpen(true);
    }

    const handleCloseModal = () => {
      setIsOpen(false);
    }

    const handlePhotoClick = (index) => {
        // console.log("enlarge func active ?")
        setEnlargedPhotoIndex(index);
        setIsEnlarged(!isEnlarged);

    }

    // console.log("let me see what is the answer looks like: ",  answer)
    return (
        <div>
            <p className="answer-text">{answer.body}</p>
            {answer.photos.map((photo, index) => (
                <img
                    key={index}
                    onClick={() => handlePhotoClick(index)}
                    src={`${answer.photos[index]}`}
                    alt="answer photo"
                    className={index === enlargePhotoIndex && isEnlarged ? 'enlarged-photo' : 'answer-photo-orign' }
                />

            ))}
             <EnlargedImageModal
                imageUrl={answer.photos[enlargePhotoIndex]}
                isOpen={isEnlarged}
                ariaHideApp={false}
                onClose={() => setIsEnlarged(false)}
            />
            <div className="answer-infor">
                <p>by {answer.answerer_name} - {new Date(answer.date).toLocaleDateString()}</p>
                {/* <button className="answer-help" >| &nbsp; Helpful? <u onClick={() => handleHelpfulClick(answer.id)}> Yes({answerHelpfulCount}) </u> </button> */}
                <button data-testid="answer-helpful-btn" className="answer-help">
                  | &nbsp; Helpful? &nbsp;
                  <u data-testid="answer-helpful-value"
                    onClick={() => {
                      if (!votedHelpful) {
                        handleHelpfulClick(answer.id);
                      }
                    }}
                    disabled={votedHelpful}
                  >
                    {votedHelpful ? 'Voted ' : `Yes (${answerHelpfulCount}) `}
                  </u>
              </button>
                {/* <button className="answer-report" onClick={handleOpenModal}>| &nbsp; <u > Report</u></button> */}
                <button data-testid="answer-report-btn" className="answer-report">
                  <p>| &nbsp; &nbsp;
                    <u
                      onClick={() => {
                        if (!reportActed) {
                          handleOpenModal()
                        }
                      }}
                      disabled={reportActed}
                      >
                        {reportActed ? 'Reported' : 'Report'}
                      </u>
                    </p>
                </button>
            </div>
            <ReactModal isOpen={isOpen} ariaHideApp={false} style={{
          content: {
            top: '50%',
            left: '50%',
            right: 'auto',
            bottom: 'auto',
            marginRight: '-50%',
            transform: 'translate(-50%, -50%)',
            width: 'auto',
            height: 'auto',
            // border-color: 'transparent',
            border: 'none',
            background: 'none'
          }
        }}>
                <button className="close-button" style={{ color: 'black', cursor: 'pointer', position: 'absolute', top: '25px', right: '25px', background:'none', border:'none' }}onClick={() => setIsOpen(false)}>[Close]</button>

                <form onSubmit={(event) => {

                    event.preventDefault();
                    // console.log("am i sending request to ans report API?", answer.id)

                    axios.put(`/qa/answers/${answer.id}/report`)
                    .then((response) => {
                          // alert("Thanks for report this answer ! ")
                          // console.log("voting succeed!")
                          setIsOpen(false);
                          setReportActed(true);
                      })
                    .catch ((error) => {
                      // alert('this answer report get error: ', error);
                     })
                    }}>
                    <div className="iron-man-form">
                      <label className="form-title">
                          Report This Answer:
                          <br />
                          <textarea
                            className="report-textarea"
                            name="reportReason"
                            onChange={(e) => setReportReason(e.target.value)}
                            style={{ height: '200px', width: '270px' }}
                            placeholder="Enter your report reason here"
                          />
                          <br/>

                          {wordCount > 4 ? <span></span> : <span className="report-word-count"> &nbsp; &nbsp; {5 - wordCount} Words To Submit</span>}

                      </label>
                      <br />
                      <div className="report-form-button-container">
                        <input
                          className="report-submit-button"
                          type="submit"
                          value="Submit"
                          disabled={wordCount < 5}
                          style={{ backgroundColor: wordCount >= 5 ? "green" : "" }}
                        />
                      </div>
                    </div>
                </form>
            </ReactModal>
        </div>
    )
}

export default AnswerEntry;
