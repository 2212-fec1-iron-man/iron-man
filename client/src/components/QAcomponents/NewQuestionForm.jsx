import React, {useState} from 'react';
import Modal from 'react-modal';
import axios from 'axios';


const NewQuestionForm = (props, closeModal) => {
  // console.log("props for the form i got :", props);
  const [questionBody, setQuestionBody] = useState('');
  const [askerName, setAskerName] = useState('');
  const [askerEmail, setAskerEmail] = useState('');

  const isValid = questionBody.length > 0 && askerName.length > 0 && /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test(askerEmail);

  const buttonClass = isValid ? "form-submit form-submit-valid" : "addquestion-form-submit";

  const handleSubmit = (event) => {
    event.preventDefault();
    const data = { body: questionBody, name: askerName, email: askerEmail, product_id:props.questionId[0]};

    // console.log("show me the data that user fill?", data);

    axios.post('http://localhost:3000/api/qa/questions', data)
    // console.log(response)#;
    .then((response) => {
      alert("Success! Your question has been submitted.");
    setTimeout(() => {
      props.closeModal();
    }, 1000);
    })
    .catch ((error) => {
      console.log(error);
    })
  }

  return (
    <form onSubmit={handleSubmit} className="iron-man-form">
      <h2 className="form-title">STARK VALUE YOUR QUESTIONS</h2>
      <div className="form-group">
        <label htmlFor="question-body" className="form-label">Question Body:</label>
        <input
          type="text"
          value={questionBody}
          onChange={e => setQuestionBody(e.target.value)}
          id="question-body"
          className="form-input"
          placeholder="Enter your question"
        />
      </div>
      <div className="form-group">
        <label htmlFor="asker-name" className="form-label">Your Name:</label>
        <input
          type="text"
          value={askerName}
          onChange={e => setAskerName(e.target.value)}
          id="asker-name"
          className="form-input"
          placeholder="Enter your name"
        />
      </div>
      <div className="form-group">
        <label htmlFor="asker-email" className="form-label">Your Email:</label>
        <input
          type="email"
          value={askerEmail}
          onChange={e => setAskerEmail(e.target.value)}
          id="asker-email"
          className="form-input"
          placeholder="Enter your email"
        />
      </div>
      <button type="submit" className={buttonClass}>Submit</button>
    </form>
  );

}
export default NewQuestionForm;