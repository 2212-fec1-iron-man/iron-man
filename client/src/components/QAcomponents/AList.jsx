import AnswerEntry from './AnswerEntry.jsx';

const AList = (props) => {
    // console.log("im in Answer render now the incoming props looks like: ", props);
    const answersArray = Object.values(props.answers);

    return (
      <div className="answer-list">
        {answersArray.map((answer, index) => (
            <AnswerEntry key={index} answer={answer} />
        ))}
      </div>
    );

}

export default AList;