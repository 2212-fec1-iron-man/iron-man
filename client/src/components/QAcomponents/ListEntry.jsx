import AList from './AList.jsx';
import {useState} from 'react';
import ReactModal from 'react-modal';
import axios from 'axios';
import styled from 'styled-components';

const BottomLine = styled.div`
background-color: #D3D3D3;
height: 2px;
width: 100%;
margin-top: 20px;
margin-bottom:15px;
`
const ListEntry = (props) => {
  // console.log("ListEntry props looks like: ", props)
  // set helpfulnessCount for showing the current question helpfulness data
  const [helpfulnessCount, setHelpfulnessCount] = useState(props.question.question_helpfulness)
  // state for voting helpful
  const [votedHelpful, setVotedHelpful] = useState(false);
  const [email, setEmail] = useState('');
  const [body, setBody] = useState('');
  const [wordCount, setWordCount] = useState(0);
  const [isValidEmail, setIsValidEmail] = useState(false);
  // handle the photos upload
  const [fileInputs, setFileInputs] = useState([0]);
  const addFileInput = () => {
    if (fileInputs.length < 3) {
      setFileInputs([...fileInputs, fileInputs.length]);
    }
  }

  const handleEmailChange = (event) => {
    const value = event.target.value;
    setEmail(value);
    const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    setIsValidEmail(re.test(value));
  }

  const handleBodyChange = (event) => {
    const value = event.target.value;
    setBody(value);
    setWordCount(value.split(" ").length);
  }

  const handleVote = (questionId) => {
    // console.log("correct question ID here?", questionId)
    axios.put(`http://localhost:3000/qa/questions/${questionId}/helpful`)
    .then((response) => {
      // alert("Thanks for voting this question helpful! ");
      // console.log(response.data);
      setVotedHelpful(true);
    })
    .catch((error) => {
      // alert("This question helpful voting got error", error);
    })
  }

  // this is handle after the click, the helpfulness accumulate
  const voteOnHelp = (questionId) => {
    // console.log("voteOnHep func, which question is dealing with right now ?", questionId);
    setHelpfulnessCount(helpfulnessCount + 1);

    handleVote(questionId);
  }
  const [isModalOpen, setIsModalOpen] = useState(false);

  // this is gonna handle the add answer button
  const addAnswer = (questionId) => {
    setIsModalOpen(!isModalOpen);
  }
  // console.log("in listEntry, i want to see the question_id :", props.question.question_id);
  // console.log("what is props here looks like?", props)
  return (
    <div >
      <div className="question-container" key={props.question.question_id}>
        <h4 className="question-body" key={props.question.question_id + 1}>Q: {props.question.question_body}</h4>
        <div key={props.question.question_id} className="question-actions" >
        <button data-testid="question-helpful-btn" className="question-help">
          Helpful? &nbsp;
          <u data-testid="question-helpful-value"
            onClick={() => {
              if (!votedHelpful) {
                voteOnHelp(props.question.question_id);
              }
            }}
            disabled={votedHelpful}
          >
            {votedHelpful ? 'Voted' : `Yes (${helpfulnessCount})`}
          </u>
        </button>
          <button key={props.question.question_id} data-testid="add-answer-btn" className="question-addAnswer" >| &nbsp; <u  onClick={() => addAnswer(props.question.question_id)}>Add Answer</u></button>
        </div>
      </div>
      <div className='flexbox-container'>
        <h4 className="answer-head">A:</h4>
        <AList answers={props.question.answers}/>
        {/* ariaHideApp is used here to prevent ReactModal fault in console */}
      </div>
      <BottomLine></BottomLine>
      <ReactModal data-testid="add-answer-modal" isOpen={isModalOpen} ariaHideApp={false} style={{
          content: {
            top: '50%',
            left: '50%',
            right: 'auto',
            bottom: 'auto',
            marginRight: '-50%',
            transform: 'translate(-50%, -50%)',
            width: 'auto',
            height: 'auto',
            border: 'none',
            background: 'none'
          }
        }}>
          <button className="close-button" style={{ color: 'black', cursor: 'pointer', position: 'absolute', top: '25px', right: '25px', background:'none', border:'none' }}onClick={() => setIsModalOpen(false)}>[Close]</button>

        <form data-testid="add-answer-modal" className="iron-man-form" onSubmit ={(event) => {
          event.preventDefault();
          let photos = [];
          for (let i = 0; i < fileInputs.length; i++) {
            if (event.target[`photos-${i}`].files.length > 0) {
              photos.push(event.target[`photos-${i}`].files[0].name);
            }
          }

          const data = {body: event.target.body.value, name: event.target.name.value, email: event.target.email.value, photos: photos}


          const addAnsUrl = `http://localhost:3000/api/qa/questions/${props.question.question_id}/answers`;

          axios.post(addAnsUrl, data)
          .then((response) => {
            // alert("Success! Your answer has been submitted.")
            setIsModalOpen(false);
          })
          .catch((error) => {
            console.log('answer submit error.')
          })
        }}>
          <div className="form-group">
            <h2 key={props.question.question_id}data-testid="add-answer-modal-header" className="form-title">Submit Your Answer ðŸ‘‡</h2>
            <textarea className="form-input" name="body" placeholder="Enter your answer here" onChange={handleBodyChange} style={{ height: '200px', width: '350px' }}></textarea>
            {wordCount >= 5 ? <span style={{ color: "green" }}> &#10003; </span> : <span className = "add-answer-body">  {5 - wordCount} words to submit </span> }
            <br/>
            <br/>
            <br/>
            {fileInputs.map((input, index) => (
                <div key={index}>
                  <input key={index} className="form-input" type="file" name={`photos-${index}`} accept="image/*" />
                  <br />
                  <br />
                </div>
              ))}
              {fileInputs.length < 3 && (
                <button data-testid="add-answer-modal-addphoto"  className="addphoto-submit" onClick={(event) => {
                  event.preventDefault();
                  addFileInput();
                }}>Add Another Photo</button>
              )}
            <br/>
            <br/>

            <input  className="form-input" type='text' name='name' placeholder='Your Name' />
            <br/>
            <br/>
            <input  className="form-input" type='email' name='email' placeholder='Your Email' onChange={handleEmailChange}/>
            {isValidEmail ? <span style={{ color: "green" }}> &#10003; </span> : <span className="add-answer-email"> Invalid Email format</span>}
            <br/>
            <br/>
            <button  data-testid="add-answer-modal-submit"   className="form-submit" type="submit" disabled={wordCount < 5 || !isValidEmail}>Add Answer</button>
          </div>
        </form>
      </ReactModal>
    </div>
  )
}
export default ListEntry;
