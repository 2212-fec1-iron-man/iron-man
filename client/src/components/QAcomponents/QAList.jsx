import ListEntry from './ListEntry.jsx';

const QAList = (props) => {
  return (
    <div>
      <br/>
      <ListEntry />
      <br/>
      <ListEntry />
      <br/>
      <ListEntry />
    </div>
  )
}

export default QAList;