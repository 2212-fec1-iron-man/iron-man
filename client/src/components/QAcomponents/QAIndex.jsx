import QAList from './QAList.jsx';
import {useState, useEffect} from 'react';
import axios from 'axios';
import NewQuestionForm from './NewQuestionForm.jsx';
import Modal from 'react-modal';

const QAIndex = () => {

  const [questionId, setQuestionId] = useState([37777])
  const [qaData, setQaData] = useState([])
  // inital state for how many questions show on the DOM
  const [questionsToShow, setQuestionsToShow] = useState(4)
  // inital state for add new question form
  const [modelIsOpen, setModelIsOpen] = useState(false)

  // below two are used for search function
  const [filteredData, setFilteredData] = useState([])
  const [searchTerm, setSearchTerm] = useState("")

  const handleSearch = () => {
    const lowerCasedSearchTerm = searchTerm.toLowerCase();
    // console.log("what is data transfer to handlesearch func",qaData[0]);
    const filtered = qaData.filter((data) => {
      return (
        data.question_body.toLowerCase().includes(lowerCasedSearchTerm)
      );
    });
    // console.log("filtered data looks like :", filtered)
    setFilteredData(filtered);
  }

  // this is for the test right now, render out questions and answers for product 37322
  useEffect(() => {

    axios.get(`http://localhost:3000/api/qa/questions/?product_id=${questionId}`)
    .then((response) => {
      // console.log("response return from server :", response)
      setQaData(response.data.results)
    })
    .catch((error) => {
      alert('first render got trouble!', error)
    })

  }, [])



  // handle the show more questions func
  function showMoreQuestions() {
    setQuestionsToShow(questionsToShow + 2);
  }

  // handle the func to add new questions
  function addNewQuestion() {
    setModelIsOpen(!modelIsOpen);
  }

  return (
    <div className="qa-wholebody">
      <p data-testid="todo-1" className="qa-head">QUESTIONS & ANSWERS</p>

      <div className="search-container">
        <input
          placeholder="HAVE A QUESTION? SEARCH FOR ANSWERS..."
          value={searchTerm}
          onChange={(e) => setSearchTerm(e.target.value)}
        />
        <button type="submit" onClick={handleSearch}>
          <img src="./magnifier.png" alt="Amplifier"/>
        </button>
      </div>

      <QAList qaData={filteredData.length > 0 ? filteredData : qaData} questionsToShow={questionsToShow}/>
      <button className="button-show-more-answered-question" onClick={showMoreQuestions}>MORE ANSWERED QUESTIONS</button>
      <button className="button-add-a-question" onClick={addNewQuestion}>ADD A QUESTION +</button>
      <Modal isOpen={modelIsOpen} onRequestClose={() => setModelIsOpen(false)} ariaHideApp={false}
        style={{
          content: {
            top: '50%',
            left: '50%',
            right: 'auto',
            bottom: 'auto',
            marginRight: '-50%',
            transform: 'translate(-50%, -50%)',
            width: 'auto',
            height: 'auto',
            border: 'none',
            background: 'none'
          }
        }}>
        <NewQuestionForm questionId={questionId} closeModal={() => setModelIsOpen(false)}/>
      </Modal>

    </div>
  )
}

export default QAIndex;
