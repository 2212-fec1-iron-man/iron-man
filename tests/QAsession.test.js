import React from 'react';
import '@testing-library/jest-dom';
import '@testing-library/react/dont-cleanup-after-each';
import userEvent from '@testing-library/user-event';
import {render, screen, waitFor, cleanup, fireEvent, waitForElement} from '@testing-library/react';


import QAIndex from '../client/src/components/QAcomponents/QAIndex.jsx';
import AList from '../client/src/components/QAcomponents/AList.jsx';
import QAList from '../client/src/components/QAcomponents/QAList.jsx';
import ListEntry from '../client/src/components/QAcomponents/ListEntry.jsx';
import AnswerEntry from '../client/src/components/QAcomponents/AnswerEntry';
import NewQuestionForm from '../client/src/components/QAcomponents/NewQuestionForm';
import axios from 'axios';

describe('QAIndex', () => {
  afterEach(cleanup);
  const user = userEvent.setup()

  it('renders QA head', () => {
    render(<QAIndex />);
    const head = screen.getByTestId('todo-1');

    // console.log("the actual head i got :",head)
    expect(head).toHaveTextContent('QUESTIONS & ANSWERS');
  });

  it ('There should be a ADD a question button on the screen', () => {

    render(<QAIndex />);
    const addQButton = screen.getByTestId("add-question-button" )
    expect(addQButton).toBeInTheDocument()
  })

  it ('There should be a More Answered question button on the screen', () => {

    render(<QAIndex />);
    const showQButton = screen.getByTestId("more-answered-questions" )
    expect(showQButton).toBeInTheDocument()
  })

  it ('There should be a question helpful button to click', () => {
    render(<ListEntry question={{answers: {2844077:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 31, question_id: 34, reported:false}} />)
    const helpfulBtn = screen.getByTestId("question-helpful-btn")
    expect(helpfulBtn).toBeInTheDocument()
  })

  it ('There should be add answer button to click', () => {
    render(<ListEntry question={{answers: {2844077:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 31, question_id: 34, reported:false}} />)
    const helpfulBtn = screen.getByTestId("add-answer-btn")
    expect(helpfulBtn).toBeInTheDocument()
  })


  it ('Each answer should have button for helpful voting ', () => {
    render(<AnswerEntry answer={{answerer_name:'tester', body: 'tester', date:'tester', helpful:1, id:1, photos:[]}}/>)
    const helpfulABtn = screen.getByTestId('answer-helpful-btn')
    expect(helpfulABtn).toBeInTheDocument()
  })

  it ('Each answer should have button for report ', () => {
    render(<AnswerEntry answer={{answerer_name:'tester', body: 'tester', date:'tester', helpful:1, id:1, photos:[]}}/>)
    const helpfulABtn = screen.getByTestId('answer-report-btn')
    expect(helpfulABtn).toBeInTheDocument()
  })

  it ('snap shots tests for AList component', () => {
    const answers = {
      1: { answerer_name: 'test1', body: 'test1', date: 'test1', helpfulness: 1, id: 1, photos: [] },
      2: { answerer_name: 'test2', body: 'test2', date: 'test2', helpfulness: 1, id: 1, photos: [] },
      3: { answerer_name: 'test3', body: 'test3', date: 'test3', helpfulness: 1, id: 1, photos: [] },
      4: { answerer_name: 'test4', body: 'test4', date: 'test4', helpfulness: 1, id: 1, photos: [] },
    };

    const { container } = render(<AList answers={answers} />);

    expect(container).toMatchSnapshot();
  })

  it ('Each answer should have button for helpful voting ', () => {
    render(<AnswerEntry answer={{answerer_name:'tester', body: 'tester', date:'tester', helpful:1, id:1, photos:[]}}/>)
    const helpfulABtn = screen.getByTestId('answer-helpful-btn')
    expect(helpfulABtn).toBeInTheDocument()
  })

  it ('answer helpful button should increase passing in helpfulness ', () => {
    render(<AnswerEntry answer={{answerer_name:'tester', body: 'tester', date:'tester', helpfulness:1, id:1, photos:[]}}/>)


    const helpfulButton = screen.getByTestId("answer-helpful-value")

    fireEvent.click(helpfulButton)

    expect(screen.getByTestId("answer-helpful-value").textContent).toBe("Yes (2) ");
  })

  it ('question helpful button should increase passing in helpfulness', () => {
    render(<ListEntry question={{answers: {2844077:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 34, reported:false}} />)


    const helpfulBtn = screen.getByTestId("question-helpful-btn")

    fireEvent.click(helpfulBtn)

    expect(screen.getByTestId("question-helpful-value").textContent).toBe("Yes (1)");
  })

  it ('snap shots tests for ListEntry component', () => {
    const question = {
      answers: {},
      asker_name: "Orval.Daugherty",
      question_body: "Recusandae nemo dolores.",
      question_date: "2021-06-22T00:00:00.000Z",
      question_helpfulness: 31,
      question_id: 304736,
      reported: false
    }

    const { container } = render(<ListEntry question={question} />);

    expect(container).toMatchSnapshot();
  })

  it('should render the form correctly', () => {
    const closeModal = jest.fn();
    const questionId = [1];
    const { getByText, getByPlaceholderText } = render(<NewQuestionForm closeModal={closeModal} questionId={questionId} />);

    expect(getByText('STARK VALUE YOUR QUESTIONS')).toBeInTheDocument();
    expect(getByText('Question Body:')).toBeInTheDocument();
    expect(getByPlaceholderText('Enter your question')).toBeInTheDocument();
    expect(getByText('Your Name:')).toBeInTheDocument();
    expect(getByPlaceholderText('Enter your name')).toBeInTheDocument();
    expect(getByText('Your Email:')).toBeInTheDocument();
    expect(getByPlaceholderText('Enter your email')).toBeInTheDocument();
    expect(getByText('Submit')).toBeInTheDocument();
  });

  it('Should increment helpful count when helpful button is clicked', () => {
    const question = {
      question_id: 1,
      question_body: 'What is your name?',
      question_helpfulness: 2,
      answers: []
    };
    const { getByTestId } = render(<ListEntry question={question} />);
    const helpfulBtn = getByTestId('question-helpful-btn');

    fireEvent.click(helpfulBtn);

    const helpfulValue = getByTestId('question-helpful-value');
    expect(helpfulValue.textContent).toBe('Yes (2)');
  });

  it ('Add answer button should pop a new window', () => {

    render(<ListEntry question={{answers: {1:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 1, reported:false}} />)
    const addAnswerButton = screen.getByText("Add Answer")
    fireEvent.click(addAnswerButton);

    expect(screen.getByTestId('add-answer-modal')).toBeInTheDocument();
  })

  it ('In the add answer pop up window, should include SUBMIT YOUR ANSWER header', () => {

    render(<ListEntry question={{answers: {1:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 1, reported:false}} />)
    const addAnswerButton = screen.getByText("Add Answer")
    fireEvent.click(addAnswerButton);

    expect(screen.getByTestId("add-answer-modal-header" )).toBeInTheDocument();
  })

  it ('In the add answer pop up window, should include add another photo button', () => {

    render(<ListEntry question={{answers: {1:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 1, reported:false}} />)
    const addAnswerButton = screen.getByText("Add Answer")
    fireEvent.click(addAnswerButton);

    expect(screen.getByTestId("add-answer-modal-addphoto")).toBeInTheDocument();
  })

  it ('In the add answer pop up window, should include submit button', () => {

    render(<ListEntry question={{answers: {1:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 1, reported:false}} />)
    const addAnswerButton = screen.getByText("Add Answer")
    fireEvent.click(addAnswerButton);

    expect(screen.getByTestId("add-answer-modal-submit")).toBeInTheDocument();
  })

  it('submits the form when all inputs are valid', async () => {
    const closeModal = jest.fn();
    const questionId = [1];

    const { getByPlaceholderText, getByText } = render(<NewQuestionForm closeModal={closeModal} questionId={questionId}/>);
    const questionInput = getByPlaceholderText('Enter your question');
    const nameInput = getByPlaceholderText('Enter your name');
    const emailInput = getByPlaceholderText('Enter your email');
    const submitButton = getByText('Submit');

    fireEvent.change(questionInput, { target: { value: 'test' } });

    fireEvent.change(nameInput, { target: { value: 'test' } });

    fireEvent.change(emailInput, { target: { value: '123@test.com' } });

    fireEvent.submit(submitButton);

    await waitFor (() => {
      expect(() => getByTestId("add-answer-modal")).toThrow();
    });
  });

  it('renders the question and answer in ListEntry', () => {
    const question = {
      question_id: 1,
      question_body: 'qtest',
      question_helpfulness: 10,
      answers: [
        {
          answer_id: 1,
          body: 'atest',
          date: '2022-01-01',
          helpfulness: 20,
          name: 'John Doe',
          photos: [],
        },
      ],
    };

    const { getByText } = render(<ListEntry question={question} />);

    expect(getByText('Q: qtest')).toBeInTheDocument();
    expect(getByText('atest')).toBeInTheDocument();
  });

  it ('snap shots tests for QAIndex component', () => {


    const { container } = render(<QAIndex />);

    expect(container).toMatchSnapshot();
  })

  it ('In the add answer pop up window, should include submit button', () => {

    render(<ListEntry question={{answers: {1:{photos:[]}}, asker_name: 'tester', question_body:'tester', question_date:'test', question_helpfulness: 1, question_id: 1, reported:false}} />)
    const addAnswerButton = screen.getByText("Add Answer")
    fireEvent.click(addAnswerButton);

    expect(screen.getByTestId("add-answer-modal-submit")).toBeInTheDocument();
  })

});