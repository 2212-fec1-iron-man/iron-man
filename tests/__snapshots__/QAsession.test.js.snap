// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QAIndex snap shots tests for AList component 1`] = `
<div>
  <div
    class="flexbox-item answer-list"
    style="display: inline-block; width: 100%; max-height: 600px; overflow-y: hidden;"
  >
    <div>
      <div>
        <p
          class="answer-text"
        >
          test1
        </p>
        <div
          class="answer-infor"
        >
          <p>
            by 
            test1
             - 
            Invalid Date
          </p>
          <button
            class="answer-help"
            data-testid="answer-helpful-btn"
          >
            | Â  Helpful? Â 
            <u
              data-testid="answer-helpful-value"
            >
              Yes (1) 
            </u>
          </button>
          <button
            class="answer-report"
            data-testid="answer-report-btn"
          >
            <p>
              | Â  Â 
              <u>
                Report
              </u>
            </p>
          </button>
        </div>
      </div>
      <div>
        <p
          class="answer-text"
        >
          test2
        </p>
        <div
          class="answer-infor"
        >
          <p>
            by 
            test2
             - 
            Invalid Date
          </p>
          <button
            class="answer-help"
            data-testid="answer-helpful-btn"
          >
            | Â  Helpful? Â 
            <u
              data-testid="answer-helpful-value"
            >
              Yes (1) 
            </u>
          </button>
          <button
            class="answer-report"
            data-testid="answer-report-btn"
          >
            <p>
              | Â  Â 
              <u>
                Report
              </u>
            </p>
          </button>
        </div>
      </div>
      <button
        class="butLoadMoreAns"
      >
        LOAD MORE ANSWERS ðŸ‘ˆ
      </button>
    </div>
  </div>
</div>
`;

exports[`QAIndex snap shots tests for ListEntry component 1`] = `
<div>
  <div>
    <div
      class="question-container"
    >
      <h4
        class="question-body"
      >
        Q: 
        Recusandae nemo dolores.
      </h4>
      <div
        class="question-actions"
      >
        <button
          class="question-help"
          data-testid="question-helpful-btn"
        >
          Helpful? Â 
          <u
            data-testid="question-helpful-value"
          >
            Yes (31)
          </u>
        </button>
        <button
          class="question-addAnswer"
          data-testid="add-answer-btn"
        >
          | Â  
          <u>
            Add Answer
          </u>
        </button>
      </div>
    </div>
    <div
      class="flexbox-container"
    >
      <h4
        class="answer-head"
      >
        A:
      </h4>
      <div
        class="flexbox-item answer-list"
        style="display: inline-block; width: 100%; max-height: 600px; overflow-y: hidden;"
      >
        <div />
      </div>
    </div>
    <div
      class="sc-bcXHqe kqoeXp"
    />
  </div>
</div>
`;
